private/protocol/eventstream/eventstreamapi/signer.go

```diff
@@ -0,0 +1,146 @@
+package eventstreamapi
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/aws/aws-sdk-go/private/protocol/eventstream"
+)
+
+const (
+	chunkAlgo            = `AWS4-HMAC-SHA256-PAYLOAD`
+	chunkSignatureHeader = ":chunk-signature"
+	chunkDateHeader      = ":date"
+	timeFormat           = "20060102T150405Z"
+)
+
+type Credentials interface {
+	Get() (credentials.Value, error)
+}
+
+type StreamingChunk interface {
+	Headers() []byte
+	Payload() []byte
+}
+
+type SigningKeyDeriver interface {
+	DeriveSigningKey(date time.Time) ([]byte, error)
+}
+
+type ScopeBuilder interface {
+	BuildScope(date time.Time) (string, error)
+}
+
+type StreamingChunkSigner struct {
+	Region  string
+	Service string
+
+	Credentials Credentials
+
+	prevSig []byte
+}
+
+func NewStreamingChunkSigner(region, service string, credentials Credentials) *StreamingChunkSigner {
+	return &StreamingChunkSigner{
+		Region:      region,
+		Service:     service,
+		Credentials: credentials,
+	}
+}
+
+func (s *StreamingChunkSigner) GetSignature(chunk StreamingChunk, date time.Time) ([]byte, error) {
```

Comment:
I think we can drop the `StreamingChunk` interface type, and use byte slices for the `header` and `payload` values directly.

```go
GetSignature(header, payload []byte, date time.Time) ([]byte, error)
```